firmware/controllers/gauges/tachometer.cpp

```diff
@@ -12,44 +12,97 @@
 
 #include "tachometer.h"
 #include "trigger_central.h"
-
-#if !EFI_UNIT_TEST
+#include "pwm_generator.h"
 
 EXTERN_ENGINE;
 
-static scheduling_s tachTurnSignalOff;
 
-static void turnTachPinLow(void *) {
-	enginePins.tachOut.setLow();
-}
+static SimplePwm tachControl("tach");
+static bool isTachEnabled;
+
+static void applyTachoPinState(int stateIndex, PwmConfig *state) /* pwm_gen_callback */ {
+	efiAssertVoid(CUSTOM_ERR_6643, stateIndex < PWM_PHASE_MAX_COUNT, "invalid stateIndex");
+	efiAssertVoid(CUSTOM_ERR_PWM_CALLBACK_ASSERT, state->multiChannelStateSequence.waveCount == 1, "invalid idle waveCount");
+	OutputPin *output = state->outputPins[0];
+	int value = state->multiChannelStateSequence.getChannelState(/*channelIndex*/0, stateIndex);
+
+	if (!value || isTachEnabled)
+		output->setValue(value);
+}        
 
 static void tachSignalCallback(trigger_event_e ckpSignalType,
```

Comment:
this doesn't need to be a trigger callback at all - call it from the period fast engine callback

Reply:
Great point.